### PR TITLE
add bot actions build, buildInRadius, buildMain

### PR DESCRIPTION
**Pull request by sweet235**
Add three bot actions:

-  `build` makes a bot choose and randomly build a buildable (similar to `roam`)
-  `buildInRadius( SOMETHING, DISTANCE )` makes a bot choose and randomly build a buildable near `SOMETHING` (similar to `roamInRadius`)
- `buildMain` makes a bot randomly build the overmind/reactor (similar to `roam`)

This is the most silly implementation of bot building I can think of. But it is already useful to help new players learn the game.

This includes an example file `build.bt`, which helps in testing this. Choose some `<bot>`, and say `/bot behavior <bot> build`. To make it behave normal again, say `/bot behavior <bot> default`.

Technical details:
- in `default.bt`, make bots spawn as dretch or with machine gun
- enable navmeshes for grangers

### CODE REVIEW

**Comment by sweet235** [comment]
Question: The bots should always build near their base. Why do we need `build` when we have `buildInRadius`? We should only use `buildInRadius( E_A_OVERMIND, 700 )`.

Answer: No. In some cases, the overmind/reactor/something is not navmesh-reachable for mysterious reasons. For an example, try map `atcshd`.

**Comment by cu-kai** [comment]
This has certainly made smaller games vs bots more fun. LGTM

**Review by slipher — COMMENTED** [review]
I think it's unnecessary to duplicate roaming logic inside the build actions. Instead we should have build actions that just attempt to build in the current location and return immediately.
```
selector {
    condition numOurBuildings(E_H_REACTOR) < 1 && numOurBuildings(E_A_OVERMIND) < 1 {
        selector {
            decorator return(STATUS_FAILURE) {
                action buildNow(E_H_REACTOR)
                action buildNow(E_A_OVERMIND)
            }
            action roam
        }
    }
    decorator return(STATUS_FAILURE) {
        condition (team == TEAM_HUMANS && distanceTo(E_H_REACTOR) < 700) ||
                  (team == TEAM_ALIENS && distanceTo(E_A_OVERMIND) < 700) {
            decorator invert {
                action buildNowChosenBuildable
            }
        }
    }
    condition(team == TEAM_ALIENS) {
        action roamInRadius(E_A_OVERMIND, 700)
    }
    condition(team == TEAM_HUMANS) {
        action roamInRadius(E_H_REACTOR, 700)
    }
    action roam
}
```
This should have more or less the same semantics as the current C++ code (note that instantaneous actions do not set the "running node", so roaming continues across build attempts as before). I'm imagining that the 'build now' actions return success iff the desired buildable is unlocked, enabled, does not exceed the max count of the buildable type (for main structure or RGS), and there is enough build points for it. `buildNow(<buildable type>)` would be similar to the `build()` function in your C++. `buildNowChosenBuildable` would use the chooseBuildableToBuild function.

But furthermore, almost all the buildable choosing logic can already be implemented in BT. The only thing missing is arithmetic operators for the `g_bot_buildNumEggs.Get() / 2` thing. 
```
selector {
    condition numOurBuildings(E_H_REACTOR) < 1 && numOurBuildings(E_A_OVERMIND) < 1 {
        selector {
            decorator return(STATUS_FAILURE) {
                action buildNow(E_H_REACTOR)
                action buildNow(E_A_OVERMIND)
            }
            action roam
        }
    }
    condition (team == TEAM_HUMANS && distanceTo(E_H_REACTOR) < 700) {
        decorator invert {
            selector {
                condition numOurBuildings(BA_H_DRILL) < 1 { action buildNow(BA_H_DRILL) }
                condition numOurBuildings(BA_H_SPAWN) < cvar("g_bot_buildNumTelenodes") { action buildNow(BA_H_SPAWN) }
                condition numOurBuildings(BA_H_ARMOURY) < 1 { action buildNow(BA_H_ARMOURY) }
                condition numOurBuildings(BA_H_MEDISTAT) < 1 { action buildNow(BA_H_MEDISTAT) }
                condition random() < cvar("g_bot_buildProbRocketpod") { action buildNow(BA_H_ROCKETPOD) }
                action buildNow(BA_H_MGTURRET)
            }
        }
    }
    condition (team == TEAM_ALIENS && distanceTo(E_A_OVERMIND) < 700) {
        decorator invert {
            selector {
                condition numOurBuildings(BA_A_LEECH) < 1 { action buildNow(BA_A_LEECH) }
                // TODO implement arithmetic
                // condition numOurBuildings(BA_A_SPAWN) < 0.5 * cvar("g_bot_buildNumEggs") {
                //    action buildNow(BA_A_SPAWN)
                // }
                condition numOurBuildings(BA_A_BOOSTER) < 1 { action buildNow(BA_A_BOOSTER) }
                condition numOurBuildings(BA_A_SPAWN) < cvar("g_bot_buildNumEggs") { action buildNow(BA_A_SPAWN) }
                action buildNow(BA_A_ACIDTUBE)
            }
        }
    }
    condition(team == TEAM_ALIENS) {
        action roamInRadius(E_A_OVERMIND, 700)
    }
    condition(team == TEAM_HUMANS) {
        action roamInRadius(E_H_REACTOR, 700)
    }
    action roam
}
```

The selector chaining for buildable selection gracefully handles the case that a buildable type is disabled, which the C++ code does not do. The one drawback of the way I wrote it above is that if, for example, a human bot has 9 BP available, it may try to to build a medistation (8 BP) rather than a telenode (10 BP), although the latter is supposed to be higher-priority.

File: pkg/unvanquished_src.dpkdir/bots/build.bt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+			{
+				condition class != PCL_ALIEN_BUILDER0 && class != PCL_ALIEN_BUILDER0_UPG
+				selector
+				{
+					sequence
+					{
+						action moveTo( E_A_OVERMIND )
+						selector
+						{
+							action evolveTo( PCL_ALIEN_BUILDER0_UPG )
```
[the remark this answers is not in the document]
Comment (sweet235) [reply pkg/unvanquished_src.dpkdir/bots/build.bt:58]: How much bigger is it?

**Comment by sweet235** [comment]
Well, when given the choice between C++ and our BT language, I'll choose C++ any day. It is a well-defined mature programming language, and a programmer can actually debug C++ programs in reasonable time. C++ programs are also faster.

Even for this most simple building scheme, the proposed changes already require adding arithmetic to the BT language. I refuse to spend my time on that. More practical building schemes will require adding more language constructs. What is our goal here? Do we really want to implement yet another scripting language?

Even if we do extend the BT language so as to make it a general purpose language, do we really want to deal with 300+ lines in BT files? Debugging a 50 line BT file is already challenging. How would we debug a 500 line BT file?

**Comment by slipher** [comment]
If you are interested in using arithmetic to the BT, I can add it real quick. If not, that's fine, you can use the first suggested BT variant where the choose building stuff is part of an action.

I do object to the current form which adds a ton of unnecessary C++ code for the roaming stuff. This type of bot AI code is brittle and requires a lot of testing to validate any change, so the duplicated roaming code will be a maintenance burden.

File: pkg/unvanquished_src.dpkdir/bots/build.bt (hunk 1 above)
[the remark this answers is not in the document]
Comment (sweet235) [reply pkg/unvanquished_src.dpkdir/bots/build.bt:58]: I am quite sure they are the same size. Compare `mins` and `maxs` in:

https://github.com/UnvanquishedAssets/res-players_src.dpkdir/blob/master/configs/classes/builder.model.cfg
https://github.com/UnvanquishedAssets/res-players_src.dpkdir/blob/master/configs/classes/builderupg.model.cfg

[the remark this answers is not in the document]
Comment (slipher) [reply pkg/unvanquished_src.dpkdir/bots/build.bt:58]: Oh good. The advanced granger is already configured to use the granger navmesh so it's fine.

**Comment by sweet235** [comment]
Thinking about this, I find that we can easily do both variants. Then we could delay the design choice until further experiments are concluded.

> I do object to the current form which adds a ton of unnecessary C++ code for the roaming stuff.

Let us look as this in detail. The largest section of duplicated code is between `BotActionRoamInRadius` and `BotActionBuildInRadius`. Removing the boilerplate present in each bot action:

```
glm::vec3 point;
botEntityAndDistance_t ent = AIEntityToGentity( self, e );

if ( !ent.ent )
{
	return STATUS_FAILURE;
}

if ( !BotFindRandomPointInRadius( self->s.number, VEC2GLM( ent.ent->s.origin ), point, radius ) )
{
	return STATUS_FAILURE;
}

if ( !BotChangeGoalPos( self, point ) )
{
	return STATUS_FAILURE;
}
```

This is just checking if `ent` is valid, and then applying two functions in sequence. The rest is handling errors. We could factor this out of `BotActionRoamInRadius` and `BotActionBuildInRadius`. But there is not much that could go wrong, as the arguments to the two called functions are given in the BT file or calculated right here.

**Comment by slipher** [comment]
> Thinking about this, I find that we can easily do both variants.

What do you mean?

Suggestions for behavior tree debugging:
- `g_bot_traceClient <clientnum>` logs the currently running action node (if any) each frame for a selected bot client.
- `action say` can be used like a debug print in the behavior tree. I have a commit #2797 to make it more convenient

**Comment by sweet235** [comment]
I could simply add an action `chooseBuildableAndBuildNow` (maybe with a nicer name), a wrapper for my C++ function `chooseBuildableAndBuild`. That would keep the potentially complicated decision logic out of the BT file. We could also count all buildable types in a single iteration.

Notice, however, that using the movement logic from `roam` and `roamInRadius` is merely a placeholder. We might come up with something better any time.

**Comment by sweet235** [comment]
> action say can be used like a debug print in the behavior tree

Not really. You have to wrap it in  `decorator ( return STATUS_FAILURE ) { ... }` or `STATUS_SUCCESS`, depending on the context. Any mistake here and your debug statement changes the execution path. I do this myself, but it is hardly practical. In C++, we can simply use `G_Say`, or any of the log functions. And all that without the pain of subtle alterations to the execution.

File: src/sgame/sg_bot_ai.cpp [hunk 2]
```diff
@@ -1188,6 +1188,220 @@ AINodeStatus_t BotActionMoveTo( gentity_t *self, AIGenericNode_t *node )
...
+AINodeStatus_t BotActionBuildMain( gentity_t *self, AIGenericNode_t *node )
+{
+	if ( !isBuilder( self ) )
+	{
+		return STATUS_FAILURE;
+	}
+
+	if ( node != self->botMind->currentNode )
+	{
+		if ( G_Team( self ) == TEAM_HUMANS && self->botMind->closestBuildings[ BA_H_REACTOR ].ent )
```
[the remark this answers is not in the document]
Comment (illwieckz) [reply src/sgame/sg_bot_ai.cpp:1373]: @sweet235 Just add a comment quoting/summarizing was what said about the too ways to do this and why you favored the one merged.

File: src/sgame/sg_bot_ai.cpp [hunk 3]
```diff
@@ -1188,6 +1188,220 @@ AINodeStatus_t BotActionMoveTo( gentity_t *self, AIGenericNode_t *node )
...
+		return false;
+	}
+
+	if ( G_Team( self ) == TEAM_HUMANS
+		 && BG_GetPlayerWeapon( &self->client->ps ) != WP_HBUILD )
+	{
+		G_ForceWeaponChange( self, WP_HBUILD );
+	}
+
+	self->client->ps.stats[ STAT_BUILDABLE ] = toBuild;
```
[the remark this answers is not in the document]
Comment (illwieckz) [reply src/sgame/sg_bot_ai.cpp:1278]: @sweet235 Just add this remark as comment for our future selves so it helps us troubleshooting if the suspicion is verified one day:

```
// slipher suspects this allows bypassing disabled buildable restrictions.
```

**Review by illwieckz — APPROVED** [review]
We need to apply “release early release often” principle to pull requests. Development is an iterative process.
If things work and doesn't do something obviously bad it's OK to go.

@sweet235 LGTM is you add the suggested comments and fix the build.

**Comment by sweet235** [comment]
All right, I condensed everything into a single bot action `buildNowChosenBuildable`, as described above.

File: src/sgame/sg_bot_ai.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (slipher) [reply src/sgame/sg_bot_ai.cpp:1286]: FWIW I think the real code executed on `/build` to see whether the structure is enabled can be reused fairly easily, like this:
```
Cmd::PushArgs("build mgturret");
Cmd_Build_f(self);
Cmd::PopArgs();
```
And then checking whether `self->client->ps.stats[ STAT_BUILDABLE ]` got changed to the selected buildable or not.

File: pkg/unvanquished_src.dpkdir/bots/build.bt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+							action evolveTo( PCL_ALIEN_BUILDER0 )
+						}
+					}
+					action suicide
+				}
+			}
+
+			decorator return ( STATUS_FAILURE )
```
Comment (slipher) [inline pkg/unvanquished_src.dpkdir/bots/build.bt:69]: I don't care one way or the other, but if you want to keep the behavior that the bot stops moving when there is nothing to build, you can use `decorator invert`

**Review by slipher — APPROVED** [review]

**Pull request merged**